Sources/web3swift/Transaction/EIP1559Envelope.swift

```diff
@@ -0,0 +1,331 @@
+    /// where Block is a block to which transaction will be included.
+    public var maxFeePerGas: BigUInt
+    public var accessList: [AccessListEntry] // from EIP-2930
+
+    private var internalChainID: BigUInt
+
+    // for CustomStringConvertible
+    public var description: String {
+        var toReturn = ""
+        toReturn += "Type: " + String(describing: self.type) + "\n"
+        toReturn += "chainID: " + String(describing: self.chainID) + "\n"
+        toReturn += "Nonce: " + String(describing: self.nonce) + "\n"
+        toReturn += "Gas limit: " + String(describing: self.gasLimit) + "\n"
+        toReturn += "Max priority fee per gas: " + String(describing: self.maxPriorityFeePerGas) + "\n"
+        toReturn += "Max fee per gas: " + String(describing: maxFeePerGas) + "\n"
+        toReturn += "To: " + self.to.address + "\n"
+        toReturn += "Value: " + String(describing: self.value) + "\n"
+        toReturn += "Data: " + self.data.toHexString().addHexPrefix().lowercased() + "\n"
+        toReturn += "Access List: " + String(describing: accessList) + "\n"
+        toReturn += "v: " + String(self.v) + "\n"
+        toReturn += "r: " + String(self.r) + "\n"
+        toReturn += "s: " + String(self.s) + "\n"
+        return toReturn
+    }
+}
+
+extension EIP1559Envelope {
+    private enum CodingKeys: String, CodingKey {
+        case chainId
+        case nonce
+        case to
+        case value
+        case maxPriorityFeePerGas
+        case maxFeePerGas
+        case gasLimit
+        case gas
+        case data
+        case input
+        case accessList
+        case v
+        case r
+        case s
+    }
+
+    public init?(from decoder: Decoder) throws {
+        let container = try decoder.container(keyedBy: CodingKeys.self)
+
+        if !container.contains(.to) || !container.contains(.nonce) || !container.contains(.value) || !container.contains(.chainId) { return nil }
+        if !container.contains(.data) && !container.contains(.input) { return nil }
+        if !container.contains(.v) || !container.contains(.r) || !container.contains(.s) { return nil }
+
+        // everything we need is present, so we should only have to throw from here
+        self.internalChainID = try container.decodeHexIfPresent(to: BigUInt.self, key: .chainId) ?? 0
+        self.nonce = try container.decodeHex(to: BigUInt.self, key: .nonce)
+
+        let list = try? container.decode([AccessListEntry].self, forKey: .accessList)
+        self.accessList = list ?? []
+
+        let toString = try? container.decode(String.self, forKey: .to)
+        switch toString {
+        case nil, "0x", "0x0":
+            self.to = EthereumAddress.contractDeploymentAddress()
+        default:
+            // the forced unwrap here is safe as we trap nil in the previous case
+            // swiftlint:disable force_unwrapping
+            guard let ethAddr = EthereumAddress(toString!) else { throw Web3Error.dataError }
```

Comment:
Just a thought: 
maybe it's worth changing `case dataError` to `case dataError(_ desc: String? = nil)`?

This way all other old use cases of `dataError` won't have to be updated right away but here and in new places we can added error messages explaining what actually happened. 

Here for example it can be:
```
Web3Error.dataError("Failed to decode \(toString!) as EthereumAddress")
```

Reply:
Same applies here:
1) https://github.com/skywinder/web3swift/pull/509/files#diff-c7bdd82e801aaff90eb009fead3875158a9d9343682ade087ea323908a723e81R93
2) https://github.com/skywinder/web3swift/pull/509/files#diff-572a2162428563df13fb402c85220f074193fc215afaa43119cabd63a05b90edR96
3) https://github.com/skywinder/web3swift/pull/509/files#diff-c7bdd82e801aaff90eb009fead3875158a9d9343682ade087ea323908a723e81R332
4) https://github.com/skywinder/web3swift/pull/509/files#diff-fa147f4d8735361ffdfe43d47471cb9fe8db5fd8bd3aa6bcc872d1269152f64fR238

Reply:
This is a great suggestion, I will leave this as a todo for another PR though, and then try to clean up the legacy cases as well. Perhaps as part of the more aggressive lint cleanup and restructuring.